cpp/src/reductions/reductions.cu

```diff
@@ -39,234 +35,160 @@ void gpu_reduction_op(const T *data, const gdf_valid_type *mask,
 
-    T agg = identity;
-
+    T_out agg = identity;
     for (int base=blkid * blksz; base<size; base+=step) {
         // Threadblock synchronous loop
         int i = base + tid;
         // load
-        T loaded = identity;
+        T_out loaded = identity;
         if (i < size && gdf_is_valid(mask, i))
-            loaded = loader(data, i);
-            
+            loaded = static_cast<T_out>(loader(data, i));
+
         // Block reduce
-        T temp = BlockReduce(temp_storage).Reduce(loaded, functor);
+        T_out temp = BlockReduce(temp_storage).Reduce(loaded, functor);
         // Add current block
         agg = functor(agg, temp);
     }
-    // First thread of each block stores the result.
-    if (tid == 0)
-        results[blkid] = agg;
-}
-
-
-
-template<typename T, typename F>
-struct ReduceOp {
-    static
-    gdf_error launch(gdf_column *input, T identity, T *output,
-                     gdf_size_type output_size) {
-        // 1st round
-        //    Partially reduce the input into *output_size* length.
-        //    Each block computes one output in *output*.
-        //    output_size == gridsize
-        typedef typename F::Loader Ld1;
-        F functor1;
-        Ld1 loader1;
-        launch_once((const T*)input->data, input->valid, input->size,
-                    (T*)output, output_size, identity, functor1, loader1);
-        CUDA_CHECK_LAST();
-
-        // 2nd round
-        //    Finish the partial reduction (if needed).
-        //    A single block reduction that computes one output stored to the
-        //    first index in *output*.
-        if ( output_size > 1 ) {
-            typedef typename F::second F2;
-            typedef typename F2::Loader Ld2;
-            F2 functor2;
-            Ld2 loader2;
-
-            launch_once(output, nullptr, output_size,
-                        output, 1, identity, functor2, loader2);
-            CUDA_CHECK_LAST();
-        }
-
-        return GDF_SUCCESS;
-    }
 
-    template <typename Functor, typename Loader>
-    static
-    void launch_once(const T *data, gdf_valid_type *valid, gdf_size_type size,
-                     T *output, gdf_size_type output_size, T identity,
-                     Functor functor, Loader loader) {
-        // find needed gridsize
-        // use atmost REDUCTION_BLOCK_SIZE blocks
-        int blocksize = REDUCTION_BLOCK_SIZE;
-        int gridsize = (output_size < REDUCTION_BLOCK_SIZE?
-                        output_size : REDUCTION_BLOCK_SIZE);
-
-        // launch kernel
-        gpu_reduction_op<<<gridsize, blocksize>>>(
-            // inputs
-            data, valid, size,
-            // output
-            output,
-            // action
-            functor,
-            // identity
-            identity,
-            // loader
-            loader
-        );
+    // First thread of each block stores the result.
+    if (tid == 0){
+        cudf::genericAtomicOperation(&results[0], agg, functor);
```

Comment:
I was on PTO when your device atomics were merged, but I just took a look and I see that ` cudf::genericAtomicOperation` will always map to an `atomicCAS`. 

However, for many operators, it would be more performant to use a more specialized atomic. For example, with the `DeviceMax` operator, I would expect `cudf::genericAtomicOperation` to use `atomicMax` when possible. The more specialized atomics will definitely be more performant than an `atomicCAS`.  An `atomicCAS` should only be used as a fallback when no other more appropriate atomic is possible.

Reply:
See https://github.com/rapidsai/cudf/issues/1377

Reply:
Looks like you already did this! Great job.